Satisfy case with multiple vision targets in frame

mvoodarla opened the pull request:
This code checks to find the two closest, largest rectangles in frame to counteract problems with the robot being confused when it sees multiple vision targets. Also, I removed some unneeded comments and shortened some of the swapping code. I also removed the side case since we won't encounter that.

shruthij28 committed:
Fix syntax errors and add checks for num of rects

Able to find parts of rectangles, distance constant is inaccurate

mvoodarla committed:
Satisfy case with multiple vision targets in frame

This code checks to find the two closest, largest rectangles in frame to
counteract problems with the robot being confused when it sees multiple
vision targets. Also, I removed some unneeded comments and shortened
some of the swapping code.

mvoodarla committed:
Calculate and return distance to vision target (#3)

Pkfire26 commented:
For distance, can we isolate the corners of the rectangle, and then compare the pixel length between two of the points with their actual dimensions to find distance to the target?

michelllepan committed:
Clean up code

Added Rectangle class to store contour properties and removed
unused code.

shruthij28 committed:
Add comments to edit certain constants

AG6GR commented on cv.py:
```diff
@@ -19,8 +19,9 @@ def __init__(self):
 
         self.FIELD_OF_VIEW_RAD = 70.42 * math.pi / 180.0
-        self.DIST_CONSTANT = 3300 # need to test this
+        self.DIST_CONSTANT = 3300 # dist_constant = focal_length * actual_height_of_reflective_tape
```
One last thing I forgot to mention: All of these distances are physical length units (mm/in etc). The size of the rectangle on your image is in pixels. There is one last conversion from pixels to physical length, which can be done with the physical sensor size of your camera. That might be hard to find online, but you can directly measure it, or design some experiments to derive it empirically.

shruthij28 committed:
Update angle and distance constants for testing

AG6GR commented:
Just a reminder, make sure to write descriptive commit messages! See https://chris.beams.io/posts/git-commit/ for guidance.

I'd recommend closing this review and starting a new one based on the commit before the reverts.

pguduguntla committed:
Use rotated boxes to represent vision targets

Added code to draw rotated boxes around vision targets and also measures distance between top two and bottom points of the boxes. Also changed the camera settings to match the port number for the jetson. Added test.py to test cv.py without having to use socket servers.

michelllepan committed:
Edit method names to follow python conventions

shruthij28 committed:
Fix distance and rename focal length constant

Changed the distance constant to be in the correct units, and also renamed ANGLE_CONST to FOcAL_LENGTH_PIXELS.

AG6GR commented on cv.py:
```diff
@@ -1,372 +1,193 @@
-        self.ANGLE_CONST = (self.SCREEN_WIDTH / 2.0) / math.tan(self.FIELD_OF_VIEW_RAD / 2.0)
 
+        self.FIELD_OF_VIEW_RAD = 70.42 * math.pi / 180.0
+        self.FOCAL_LENGTH_PIXELS = (self.SCREEN_WIDTH / 2.0) / math.tan(self.FIELD_OF_VIEW_RAD / 2.0) #angle constant is the same as focal length of camera (pixels)
```
Can you update the comment? Ideally explain this formula.

michelllepan replied:
resolved by b0cf155

AG6GR commented on cv.py:
```diff
@@ -1,372 +1,193 @@
-        y = (y1 + y2 + y3 + y4) / 4.0;
+    def calc_dist(self, length):
+        if(length > 0):
+            return (self.FOCAL_LENGTH_PIXELS * 5.5) / length
```
What's 5.5? How did you find this number?

AG6GR commented on cv.py:
```diff
@@ -1,372 +1,193 @@
         mask = cv2.inRange(hsv, low_green, high_green)
+        _, contours,_ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+
+        def sort_by_area(c):
```
You should be able to pass in cv2.contourArea directly, no need to create a dummy function.

michelllepan replied:
resolved by b0cf155

AG6GR commented on cv.py:
```diff
@@ -1,372 +1,193 @@
+            print rotated_rect2.box
+            print rotated_rect2.rot_angle
+
+            cv2.drawContours(frame,[rotated_rect1.box],0,(0,0,255),2)
```
Spaces after commas please. Just like regular English.

michelllepan replied:
resolved by b0cf155

AG6GR commented on cv.py:
```diff
@@ -1,372 +1,193 @@
+            print "bottom dist: " + str(bottom_dist)
+
+
+            #i dont know how the points work
```
Can we change this comment to explain how the points work? Also next time if you don't understand something, feel free to ask on Slack. We're all learning together, it's ok to have questions.

michelllepan replied:
resolved by b0cf155

AG6GR commented on cv.py:
```diff
@@ -1,372 +1,193 @@
+class RotatedRectangle:
 
+    def __init__(self, box, area, rot_angle):
+        self.box = box
```
It might be helpful to comment what type you expect some of these variables to be. Especially if they are an object rather than a simple type.

michelllepan replied:
resolved by b0cf155

AG6GR commented on run.py:
```diff
@@ -17,6 +17,6 @@
-    print str(angle)
-    conn.sendall(str(angle)+'\n')
+    angle, distance = detector.run_cv()
+    print str(angle), str(distance)
```
Is there a reason we aren't using Python3?

For bonus points, take a look at https://pyformat.info/ . format() is a really awesome feature of Python that makes printing stuff like this much cleaner.

mvoodarla replied:
We honestly should use Python3.

AG6GR replied:
Let's discuss this at a future meeting. It isn't exactly a trivial change since we would have to set up OpenCV again for everyone in Python 3.

AG6GR commented on test.py:
```diff
@@ -0,0 +1,10 @@
+detector = VisionTargetDetector()
+
+while True:
+    time.sleep(0.1)
```
Just curious what happens if you don't do time.sleep()?

AG6GR reviewed (changes requested):
Some minor cleanup changes. I didn't go over the code logic in detail, we can talk about that tomorrow.

For code style, I'm going off the [PEP8](https://www.python.org/dev/peps/pep-0008/) Python standard.

michelllepan committed:
Add comments and remove extra print statements

michelllepan committed:
Fix indenting

michelllepan committed:
Change text size and color

shruthij28 committed:
Change exposure in camerasettings.sh
